### DISCUSSION TITLE
Saleor Provider getting customized Apollo config and Cache Rehydration

### DISCUSSION THREAD

**Discussion by TurkaljGoran** [saleor/saleor#8351] [Q&A]
Hello, 

I have a few related questions about the lastest Saleor SDK: 

1. How can my custom Apollo configuration be passed into the Saleor Provider along with saleorClient? Previouse version of SDK did it in a way that you could pass config="saleorConfig" and apolloConfig="customApolloConfig":

Example from NPM for SDK 0.1.5: 

import { SaleorProvider, useAuth } from "@saleor/sdk";
 
const config = { apiUrl: "http://localhost:8000/graphql/" };
const apolloConfig = {
  /* 
    Optional custom Apollo client config.
    Here you may append custom Apollo cache, links or the whole client. 
    You may also use import { createSaleorCache, createSaleorClient, createSaleorLinks } from "@saleor/sdk" to create semi-custom implementation of Apollo.
  */
};
 
const rootElement = document.getElementById("root");
ReactDOM.render(
  <SaleorProvider config={config} apolloConfig={apolloConfig}>
    <App />
  </SaleorProvider>,
  rootElement
);

- The above does not work any longer with the 0.4.0-4 version of SDK as far as I can see because you can only pass 'client' prop to SaleorProvider whic takes saleorClient object. 
- Is there a way to use both custom Apollo client and saleorClient together?

Currently I'm doing this - using the private apollo client on the saleorClient itself and passing it to <ApolloProvider> which wraps everything and it does work but feels like horrible anti-pattern:

 <ApolloProvider client={saleorClient._internal.apolloClient}>
          <SaleorProvider client={saleorClient}>
            <DefaultLayout>
                   <Component {...pageProps} /> 
              </AnimatePresence>
            </DefaultLayout>
          </SaleorProvider>
        </ApolloProvider>

2. Is this the only way to go for in order to NOT have two separate Apollo clients (one custom and one saleorClient independently)?

3. Further more, if I access saleorClient.internal.apolloClient in one of Next JS's data fetching methdos like getServerSideProps or getStaticProps it works ok - but does it need cache extraction so that when the query is made during SSR, that the cache from the server side apollo client (saleorClient._internal.apolloClient in this case) can be sent back to the client side and merged with client side apollo cache so that the query WONT be made again because data is already in the cache ?

Is this maybe done 'under' the hood somehow with the SDK?

This is an example of what I mean by server side cache extraction from the apollo client: 

![extract example](https://user-images.githubusercontent.com/75638667/137079023-43746572-f469-4d0e-806e-77f19103af4a.jpg)

**Comment by Diego17ccs** [comment]
![2021-10-13](https://user-images.githubusercontent.com/90481667/137237334-b5f40071-80d1-4f25-9585-fbae89515c9f.png)
